text/0000-subslice-pattern-syntax.md

```diff
@@ -0,0 +1,207 @@
+
+```rust
+// Binding a sub-array:
+let [x, y @ .., z] = [1, 2, 3, 4]; // `y: [i32, 2] = [2, 3]`
+
+// Binding a sub-slice:
+let [x, y @ .., z]: &[u8] = &[1, 2, 3, 4]; // `y: &[i32] = &[2, 3]`
+```
+
+# Motivation
+[motivation]: #motivation
+
+## General motivation
+Stabilization of slice pattern with subslices is currently blocked on finalizing syntax for
+these subslices.  
+This RFC proposes a syntax for stabilization.
+
+## Motivation for the specific syntax
+
+### The shortcut form: `..`
+
+This form is already used in the meaning "rest of the list" in struct patterns, tuple struct
+patterns and tuple patterns so it would be logical to use it for slice patterns as well.  
+And indeed, in unstable Rust `..` is used in this meaning since long before 1.0.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Sub-slices and sub-arrays can be matched using `..` and `<IDENT> @ ..` can be used to bind
+these sub-slices and sub-arrays to an identifier.
+
+```rust
+// Matching slices using `ref` patterns:
+let v = vec![1, 2, 3];
+match v[..] {
+    [1, ref subslice @ .., 4] => assert_eq!(subslice.len(), 1),
+    [5, ref subslice @ ..] => assert_eq!(subslice.len(), 2),
+    [ref subslice @ .., 6] => assert_eq!(subslice.len(), 2),
+    [x, .., y] => assert!(v.len() >= 2),
+    [..] => {} // Always matches
+}
+
+// Matching slices using default-binding-modes:
+let v = vec![1, 2, 3];
+match &v[..] {
+    [1, subslice @ .., 4] => assert_eq!(subslice.len(), 1),
+    [5, subslice @ ..] => assert_eq!(subslice.len(), 2),
+    [subslice @ .., 6] => assert_eq!(subslice.len(), 2),
+    [x, .., y] => assert!(v.len() >= 2),
+    [..] => {} // Always matches
+}
+
+// Matching arrays by-value:
+let v = [1, 2, 3];
+match v {
+  [1, subarray @ .., 3] => assert_eq!(subarray, [2]),
+  [5, subarray @ ..] => has_type::<[i32; 2]>(subarray),
+  [subarray @ .., 6] => has_type::<[i32, 2]>(subarray),
+  [x, .., y] => has_type::<[i32, 1]>(x),
+  [..] => {},
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`..` can be used as a pattern fragment for matching sub-slices and sub-arrays.
+It is treated as a "non-reference-pattern" for the purpose of determining default-binding-modes,
+and so shifts the binding mode to by-`ref` or by-`ref mut` when used to match a subsection of a
+reference or mutable reference to a slice or array.
+
+`@` can be used to bind the result of `..` to an identifier.
+
+ When used to match against a non-reference slice (`[u8]`), `x @ ..` would attempt to bind
+by-value, which would fail in the case that users haven't enabled `feature(unsized_locals)`
+(since otherwise it's not possible to bind `[u8]` to a variable directly).
+
+`..`/`IDENT @ ..` is not a full pattern syntax, but rather a part of slice, tuple and tuple
+struct pattern syntaxes. In particular, `..` is not accepted by the `pat` macro matcher.
```

Comment:
So then it is not clear to me what the syntax is... ISTM that we have something like:
```rust
Pat =
  | // everything else in the pattern grammar...
  | Slice:{ "[" elems:SubPat* % "," ","? "]" }
  | Tuple:{ "(" fields:SubPat* % "," ","? ")" }
  | TupleStruct:{ path:Path "(" fields:SubPat* % "," ","? ")" }
  ;

Ref = "ref" mut:"mut"?;
SlicePatBinding = ref:Ref? ident:IDENT;
SlicePat = {binding:SlicePatBinding "@"}? "..";

SubPat =
  | Slice:SlicePat
  | Pat:Pat
  ;
```
Is that correct? ISTM that it would be cleaner to make `..` into a full pattern and then reject what we don't want semantically. If we allow `..` to be a full pattern, macro authors can also make use of `..` for their own DSL purposes.

If correct, you should also give semantics (including what the types are) for `{ref mut?}? x @ ..` in tuples and tuple structs or specify that they are semantically rejected.

Reply:
The goal is to keep status quo and make only changes necessary for stabilization of slice patterns.
`..` is a fragment of the slice, tuple, tuple struct and struct pattern syntaxes.
`ref? mut? IDENT @ ..` is a fragment of the slice pattern syntax only.
Duplicate `..`s are also rejected at syntactic level, there are no semantic errors.

Reply:
>The RFC states that let (x, y @ ..) = expr; and let Struct(x, y @ ..) = expr; are allowed syntactically

Where?
If something looks like it states that, then it needs to be fixed.

Reply:
Line 89; 
> `..`/`IDENT @ ..` is not a full pattern syntax, but rather a part of slice, tuple and tuple
struct pattern syntaxes.

From this follows that `(x, y @ ..)` is legal or at best it is ambiguous.